src/Processors/Transforms/Streaming/JoinTransformWithAlignment.cpp

```diff
@@ -180,121 +150,368 @@ IProcessor::Status JoinTransformWithAlignment::prepare()
-///    b. If events' left_stream_watermark + latency_threshold > right_stream_watermark, buffer these events in JoinTransform.
+///       ii) Continue pulling more left stream data until the `left_stream_watermark + latency_threshold >= right_stream_watermark`.
+///            and buffer these events which hold `left_stream_watermark + latency_threshold >= right_stream_watermark` events in JoinTransform for next join.
+///    b. If events' left_stream_watermark + latency_threshold >= right_stream_watermark, buffer these events in JoinTransform.
 ///       Mute left stream and unmute right stream if necessary
 /// 3) If there are already buffered data for left stream (it shall be muted). Check if we can send the buffered to join the right hashtable and send as
 ///    much as possible. If the left buffered data is empty, unmute it.
 ///
-/// If left input got stuck, basically we will need mute right input to avoid having too much memory pressure.
-/// If right input got stuck, we wait for quiesce_threshold period, after that, we continue the join which may join stale right data
+/// If left input got stuck, basically we will need mute right input to avoid having too much memory pressure or wait for quiesce_threshold period.
+/// If right input got stuck, we will need mute left input to avoid having too much memory pressure or wait for quiesce_threshold period.
+/// After that, we continue the join which may join stale left/right data
 /// but this is the best we can do since if we don't join, we stuck forever (this is a trade off, we can choose to stuck until right input
 /// has more data, but in some scenario, they just didn't have more data).
-///
-/// When right stream can garbage collect its data ?
-/// For asof join, for now, we manually use keep_versions to do garbage collection. Ideally we can do better job (automatically), but seems hard since version
-/// and watermark / timestamp are different.
+
 void JoinTransformWithAlignment::work()
 {
+    bool left_input_in_quiesce = isInputInQuiesce(left_input);
+    bool right_input_in_quiesce = isInputInQuiesce(right_input);
+
+    /// Right input has buffered data or newly pulled data
+    while (right_input.hasCompleteChunks())
     {
-        std::scoped_lock lock(mutex);
+        if (isCancelled())
+            return;
 
-        /// 1) We pulled right input and it has data
-        if (right_input.input_chunk)
+        auto & [retracted_chunk, chunk] = right_input.input_chunks.front();
+
+        /// We can directly process current input data in followsing scenarios:
+        /// 1) Don't need align inputs (right side of join has builtin buffer to align)
+        /// 2) Right input chunk's max timestamp less than left stream's min timestamp
+        /// 3) Or left input is in quiesce, we still need to push execute the join probably
+        if (!right_input.need_aligned_buffer || chunk.maxTimestamp() <= left_input.minTimestamp() + latency_threshold
```

Comment:
I think we shall compare watermark here, left watermark, right watermark and the latency threshold ?

Reply:
Compare watermark is great, it's one alignment mode, as you said, I will do more alignment modes in next ticket